src/server/transaction.cc

```diff
@@ -338,7 +337,7 @@ bool Transaction::RunInShard(EngineShard* shard) {
     DCHECK(!awaked_prerun);  // we should not have blocking transaction inside multi block.
 
     sd.local_mask |= KEYLOCK_ACQUIRED;
-    shard->db_slice().Acquire(mode, GetLockArgs(idx));
+    shard->db_slice().Acquire(mode, GetLockArgs(shard->shard_id()));
```

Comment:
I need to see why it's correct. Specifically, we keep all the data in the 0 index for single-shard transactions.
I am guessing you move this logic to GetLockArgs but right now I do not see how I can review all the reorderings and functionality changes AND keep my sanity.